test/hotspot/jtreg/runtime/cds/appcds/aotCache/JcmdAOTEndTrainingTest.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2021, Red Hat, Inc. All rights reserved.
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test
+ * @requires vm.cds.supports.aot.class.linking
+ * @requires vm.cds.write.archived.java.heap
+ * @summary Sanity test for Jcmd AOT.end_training command
+ * @library /test/lib
+ * @build JcmdAOTEndTrainingTest
+ * @run driver JcmdAOTEndTrainingTest
+ */
+
+import jdk.test.lib.JDKToolLauncher;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.apps.LingeredApp;
+import jdk.test.lib.dcmd.PidJcmdExecutor;
+import jdk.test.lib.process.OutputAnalyzer;
+import java.io.IOException;
+
+public class JcmdAOTEndTrainingTest {
+    public static void main(String[] args)  throws Exception {
+        LingeredApp theApp = null;
+        try {
+            theApp = new LingeredApp();
+            theApp.setUseDefaultClasspath(false);
+            LingeredApp.startApp(theApp);
+            long pid = theApp.getPid();
+
+            JDKToolLauncher jcmd = JDKToolLauncher.createUsingTestJDK("jcmd");
+            jcmd.addToolArg(String.valueOf(pid));
+            jcmd.addToolArg("AOT.end_training");
+
+            try {
+                OutputAnalyzer output = ProcessTools.executeProcess(jcmd.getCommand());
+                output.shouldContain("Error! Not a training run");
```

Comment:
So this test only checks to see that we're *not* in a training run? Can we add other tests, e.g., ending an actual training run?

Reply:
This test basically checks that the JCmd has been compiled; it proved difficult to create an AOT test (recording) and run the JCmd from the same test.  Now I have basic tests done I'll go back and address the recording tests